File: cardgrader/transform.py
"""Mapping between image space and canvas (view) space.

    view = pan + scale * R(theta) * (p_img - image_center)

R is expressed in screen coordinates (y down), so a positive angle rotates the
card clockwise on screen.  The same transform is handed to PIL as a single
affine so rotation, scaling and panning cost one resample per frame.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

import numpy as np

from .geometry import vec


@dataclass
class ViewTransform:
    angle: float = 0.0          # degrees, positive = clockwise on screen
    scale: float = 1.0
    pan: np.ndarray = field(default_factory=lambda: vec(0.0, 0.0))   # canvas px
    center: np.ndarray = field(default_factory=lambda: vec(0.0, 0.0))  # image px

    # --- helpers ----------------------------------------------------------
    def _cos_sin(self):
        r = math.radians(self.angle)
        return math.cos(r), math.sin(r)

    def img_to_view(self, p) -> np.ndarray:
        c, s = self._cos_sin()
        d = np.asarray(p, dtype=float) - self.center
        return vec(
            self.pan[0] + self.scale * (c * d[0] - s * d[1]),
            self.pan[1] + self.scale * (s * d[0] + c * d[1]),
        )

    def view_to_img(self, p) -> np.ndarray:
        c, s = self._cos_sin()
        d = np.asarray(p, dtype=float) - self.pan
        return vec(
            self.center[0] + (c * d[0] + s * d[1]) / self.scale,
            self.center[1] + (-s * d[0] + c * d[1]) / self.scale,
        )

    def view_delta_to_img(self, d) -> np.ndarray:
        """Rotate/scale a *delta* (no translation) from view into image space."""
        c, s = self._cos_sin()
        d = np.asarray(d, dtype=float)
        return vec((c * d[0] + s * d[1]) / self.scale, (-s * d[0] + c * d[1]) / self.scale)

    def affine_coeffs(self, scale=None, pan=None):
        """Coefficients for PIL Image.transform(AFFINE): out -> in."""
        sc = self.scale if scale is None else scale
        pn = self.pan if pan is None else np.asarray(pan, dtype=float)
        c, s = self._cos_sin()
        a, b = c / sc, s / sc
        d, e = -s / sc, c / sc
        cc = self.center[0] - (a * pn[0] + b * pn[1])
        ff = self.center[1] - (d * pn[0] + e * pn[1])
        return (a, b, cc, d, e, ff)

    def fit(self, img_size, canvas_size, margin=0.90) -> None:
        iw, ih = img_size
        cw, ch = canvas_size
        self.center = vec(iw / 2.0, ih / 2.0)
        if iw <= 0 or ih <= 0 or cw <= 0 or ch <= 0:
            return
        self.scale = min(cw / iw, ch / ih) * margin
        self.pan = vec(cw / 2.0, ch / 2.0)

    def zoom_about(self, view_point, factor: float, lo=0.02, hi=40.0) -> None:
        """Zoom keeping the image point under `view_point` stationary."""
        anchor = self.view_to_img(view_point)
        self.scale = max(lo, min(hi, self.scale * factor))
        new_view = self.img_to_view(anchor)
        self.pan = self.pan + (np.asarray(view_point, dtype=float) - new_view)

    def rotate_about(self, view_point, delta_deg: float) -> None:
        anchor = self.view_to_img(view_point)
        self.angle = (self.angle + delta_deg + 180.0) % 360.0 - 180.0
        new_view = self.img_to_view(anchor)
        self.pan = self.pan + (np.asarray(view_point, dtype=float) - new_view)

File: src/core/transform.ts
/**
 * Mapping between image space and view (canvas) space.
 *
 *     view = pan + scale * R(angle) * (p_img - center)
 *
 * R is expressed in screen coordinates (y down), so a positive angle rotates
 * the card clockwise on screen.
 */

import { type Vec, pymod } from "./geometry";

export class ViewTransform {
  constructor(
    public angle = 0, // degrees, positive = clockwise on screen
    public scale = 1,
    public pan: Vec = [0, 0], // view px
    public center: Vec = [0, 0], // image px
  ) {}

  copy(): ViewTransform {
    return new ViewTransform(this.angle, this.scale, [...this.pan], [...this.center]);
  }

  private cs(): [number, number] {
    const r = (this.angle * Math.PI) / 180;
    return [Math.cos(r), Math.sin(r)];
  }

  imgToView(p: Vec): Vec {
    const [c, s] = this.cs();
    const dx = p[0] - this.center[0];
    const dy = p[1] - this.center[1];
    return [
      this.pan[0] + this.scale * (c * dx - s * dy),
      this.pan[1] + this.scale * (s * dx + c * dy),
    ];
  }

  viewToImg(p: Vec): Vec {
    const [c, s] = this.cs();
    const dx = p[0] - this.pan[0];
    const dy = p[1] - this.pan[1];
    return [
      this.center[0] + (c * dx + s * dy) / this.scale,
      this.center[1] + (-s * dx + c * dy) / this.scale,
    ];
  }

  /** Rotate and scale a *delta* (no translation) from view into image space. */
  viewDeltaToImg(d: Vec): Vec {
    const [c, s] = this.cs();
    return [(c * d[0] + s * d[1]) / this.scale, (-s * d[0] + c * d[1]) / this.scale];
  }

  /**
   * The same mapping as a canvas matrix (image -> view), for
   * `ctx.setTransform(a, b, c, d, e, f)`.
   */
  canvasMatrix(): [number, number, number, number, number, number] {
    const [c, s] = this.cs();
    const k = this.scale;
    const a = k * c;
    const b = k * s;
    const cc = -k * s;
    const d = k * c;
    const e = this.pan[0] - (a * this.center[0] + cc * this.center[1]);
    const f = this.pan[1] - (b * this.center[0] + d * this.center[1]);
    return [a, b, cc, d, e, f];
  }

  fit(imgSize: Vec, canvasSize: Vec, margin = 0.9): void {
    const [iw, ih] = imgSize;
    const [cw, ch] = canvasSize;
    this.center = [iw / 2, ih / 2];
    if (iw <= 0 || ih <= 0 || cw <= 0 || ch <= 0) return;
    this.scale = Math.min(cw / iw, ch / ih) * margin;
    this.pan = [cw / 2, ch / 2];
  }

  /** Zoom keeping the image point under `viewPoint` stationary. */
  zoomAbout(viewPoint: Vec, factor: number, lo = 0.02, hi = 40): void {
    const anchor = this.viewToImg(viewPoint);
    this.scale = Math.max(lo, Math.min(hi, this.scale * factor));
    const nv = this.imgToView(anchor);
    this.pan = [this.pan[0] + (viewPoint[0] - nv[0]), this.pan[1] + (viewPoint[1] - nv[1])];
  }

  rotateAbout(viewPoint: Vec, deltaDeg: number): void {
    const anchor = this.viewToImg(viewPoint);
    this.angle = pymod(this.angle + deltaDeg + 180, 360) - 180;
    const nv = this.imgToView(anchor);
    this.pan = [this.pan[0] + (viewPoint[0] - nv[0]), this.pan[1] + (viewPoint[1] - nv[1])];
  }
}
